core/server/common/src/main/java/alluxio/master/journal/raft/SnapshotReplicationManager.java

```diff
@@ -287,7 +295,11 @@ public long maybeCopySnapshotFromFollower() {
           return termIndex;
         }).exceptionally(e -> {
           LOG.error("Unexpected exception downloading snapshot from follower {}.", followerIp, e);
-          transitionState(DownloadState.STREAM_DATA, DownloadState.IDLE);
+          // this allows the leading master to request other followers for their snapshots. It
+          // previously collected information about other snapshots in requestInfo(). If no other
+          // snapshots are available requestData() will return false and mDownloadState will be IDLE
+          transitionState(DownloadState.STREAM_DATA, DownloadState.REQUEST_DATA);
+          CompletableFuture.runAsync(this::requestSnapshotFromFollowers);
```

Comment:
wouldn't this happen automatically if you had just set the state to IDLE?
by automatic I mean, doesn't `JournalStateMachine` already keeps invoking the snapshot request code?

Reply:
If we reset to `IDLE` we may run into the same problem we had before: the follower with the best snapshot is able to communicate which snapshot it has, but unable to upload it (for example, if the network is flaky, a simple request for term and index might get through but not a whole snapshot upload).
Going back to `REQUEST_DATA` allows other followers to attempt to upload their snapshots, from best snapshot to worst snapshot.